// @paperback/toolchain/dist/devtools/generated/typescript/PDTLogger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PaperbackLogger = exports.LogFilter = exports.LogLine = exports.LogLevel = void 0;
// @generated by protobuf-ts 2.8.3 with parameter client_grpc1
// @generated from protobuf file "PDTLogger.proto" (syntax proto3)
// tslint:disable
const runtime_rpc_1 = require("@protobuf-ts/runtime-rpc");
const runtime_1 = require("@protobuf-ts/runtime");
const runtime_2 = require("@protobuf-ts/runtime");
const runtime_3 = require("@protobuf-ts/runtime");
const runtime_4 = require("@protobuf-ts/runtime");
const runtime_5 = require("@protobuf-ts/runtime");
const timestamp_1 = require("./google/protobuf/timestamp");
/**
 * @generated from protobuf enum LogLevel
 */
var LogLevel;
(function (LogLevel) {
    /**
     * @generated from protobuf enum value: INFO = 0;
     */
    LogLevel[LogLevel["INFO"] = 0] = "INFO";
    /**
     * @generated from protobuf enum value: WARN = 1;
     */
    LogLevel[LogLevel["WARN"] = 1] = "WARN";
    /**
     * @generated from protobuf enum value: ERROR = 2;
     */
    LogLevel[LogLevel["ERROR"] = 2] = "ERROR";
})(LogLevel = exports.LogLevel || (exports.LogLevel = {}));
// @generated message type with reflection information, may provide speed optimized methods
class LogLine$Type extends runtime_5.MessageType {
    constructor() {
        super("LogLine", [
            { no: 1, name: "level", kind: "enum", T: () => ["LogLevel", LogLevel] },
            { no: 2, name: "tags", kind: "scalar", repeat: 2 /*RepeatType.UNPACKED*/, T: 9 /*ScalarType.STRING*/ },
            { no: 3, name: "message", kind: "scalar", T: 9 /*ScalarType.STRING*/ },
            { no: 4, name: "date", kind: "message", T: () => timestamp_1.Timestamp }
        ]);
    }
    create(value) {
        const message = { level: 0, tags: [], message: "" };
        globalThis.Object.defineProperty(message, runtime_4.MESSAGE_TYPE, { enumerable: false, value: this });
        if (value !== undefined)
            (0, runtime_3.reflectionMergePartial)(this, message, value);
        return message;
    }
    internalBinaryRead(reader, length, options, target) {
        let message = target ?? this.create(), end = reader.pos + length;
        while (reader.pos < end) {
            let [fieldNo, wireType] = reader.tag();
            switch (fieldNo) {
                case /* LogLevel level */ 1:
                    message.level = reader.int32();
                    break;
                case /* repeated string tags */ 2:
                    message.tags.push(reader.string());
                    break;
                case /* string message */ 3:
                    message.message = reader.string();
                    break;
                case /* google.protobuf.Timestamp date */ 4:
                    message.date = timestamp_1.Timestamp.internalBinaryRead(reader, reader.uint32(), options, message.date);
                    break;
                default:
                    let u = options.readUnknownField;
                    if (u === "throw")
                        throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
                    let d = reader.skip(wireType);
                    if (u !== false)
                        (u === true ? runtime_2.UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
            }
        }
        return message;
    }
    internalBinaryWrite(message, writer, options) {
        /* LogLevel level = 1; */
        if (message.level !== 0)
            writer.tag(1, runtime_1.WireType.Varint).int32(message.level);
        /* repeated string tags = 2; */
        for (let i = 0; i < message.tags.length; i++)
            writer.tag(2, runtime_1.WireType.LengthDelimited).string(message.tags[i]);
        /* string message = 3; */
        if (message.message !== "")
            writer.tag(3, runtime_1.WireType.LengthDelimited).string(message.message);
        /* google.protobuf.Timestamp date = 4; */
        if (message.date)
            timestamp_1.Timestamp.internalBinaryWrite(message.date, writer.tag(4, runtime_1.WireType.LengthDelimited).fork(), options).join();
        let u = options.writeUnknownFields;
        if (u !== false)
            (u == true ? runtime_2.UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
        return writer;
    }
}
/**
 * @generated MessageType for protobuf message LogLine
 */
exports.LogLine = new LogLine$Type();
// @generated message type with reflection information, may provide speed optimized methods
class LogFilter$Type extends runtime_5.MessageType {
    constructor() {
        super("LogFilter", []);
    }
    create(value) {
        const message = {};
        globalThis.Object.defineProperty(message, runtime_4.MESSAGE_TYPE, { enumerable: false, value: this });
        if (value !== undefined)
            (0, runtime_3.reflectionMergePartial)(this, message, value);
        return message;
    }
    internalBinaryRead(reader, length, options, target) {
        return target ?? this.create();
    }
    internalBinaryWrite(message, writer, options) {
        let u = options.writeUnknownFields;
        if (u !== false)
            (u == true ? runtime_2.UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
        return writer;
    }
}
/**
 * @generated MessageType for protobuf message LogFilter
 */
exports.LogFilter = new LogFilter$Type();
/**
 * @generated ServiceType for protobuf service PaperbackLogger
 */
exports.PaperbackLogger = new runtime_rpc_1.ServiceType("PaperbackLogger", [
    { name: "streamLogs", serverStreaming: true, options: {}, I: exports.LogFilter, O: exports.LogLine }
]);
