// @paperback/toolchain/dist/devtools/generated/typescript/PDTSourceTester.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PaperbackSourceTester = exports.SourceTestResponse = exports.SourceTestRequest = exports.TestData = exports.SearchData = exports.SourceInstallResponse = exports.SourceInstallRequest = void 0;
// @generated by protobuf-ts 2.8.3 with parameter client_grpc1
// @generated from protobuf file "PDTSourceTester.proto" (syntax proto3)
// tslint:disable
const runtime_rpc_1 = require("@protobuf-ts/runtime-rpc");
const runtime_1 = require("@protobuf-ts/runtime");
const runtime_2 = require("@protobuf-ts/runtime");
const runtime_3 = require("@protobuf-ts/runtime");
const runtime_4 = require("@protobuf-ts/runtime");
const runtime_5 = require("@protobuf-ts/runtime");
const timestamp_1 = require("./google/protobuf/timestamp");
// @generated message type with reflection information, may provide speed optimized methods
class SourceInstallRequest$Type extends runtime_5.MessageType {
    constructor() {
        super("SourceInstallRequest", [
            { no: 1, name: "sourceId", kind: "scalar", T: 9 /*ScalarType.STRING*/ },
            { no: 2, name: "repoBaseUrl", kind: "scalar", T: 9 /*ScalarType.STRING*/ }
        ]);
    }
    create(value) {
        const message = { sourceId: "", repoBaseUrl: "" };
        globalThis.Object.defineProperty(message, runtime_4.MESSAGE_TYPE, { enumerable: false, value: this });
        if (value !== undefined)
            (0, runtime_3.reflectionMergePartial)(this, message, value);
        return message;
    }
    internalBinaryRead(reader, length, options, target) {
        let message = target ?? this.create(), end = reader.pos + length;
        while (reader.pos < end) {
            let [fieldNo, wireType] = reader.tag();
            switch (fieldNo) {
                case /* string sourceId */ 1:
                    message.sourceId = reader.string();
                    break;
                case /* string repoBaseUrl */ 2:
                    message.repoBaseUrl = reader.string();
                    break;
                default:
                    let u = options.readUnknownField;
                    if (u === "throw")
                        throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
                    let d = reader.skip(wireType);
                    if (u !== false)
                        (u === true ? runtime_2.UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
            }
        }
        return message;
    }
    internalBinaryWrite(message, writer, options) {
        /* string sourceId = 1; */
        if (message.sourceId !== "")
            writer.tag(1, runtime_1.WireType.LengthDelimited).string(message.sourceId);
        /* string repoBaseUrl = 2; */
        if (message.repoBaseUrl !== "")
            writer.tag(2, runtime_1.WireType.LengthDelimited).string(message.repoBaseUrl);
        let u = options.writeUnknownFields;
        if (u !== false)
            (u == true ? runtime_2.UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
        return writer;
    }
}
/**
 * @generated MessageType for protobuf message SourceInstallRequest
 */
exports.SourceInstallRequest = new SourceInstallRequest$Type();
// @generated message type with reflection information, may provide speed optimized methods
class SourceInstallResponse$Type extends runtime_5.MessageType {
    constructor() {
        super("SourceInstallResponse", []);
    }
    create(value) {
        const message = {};
        globalThis.Object.defineProperty(message, runtime_4.MESSAGE_TYPE, { enumerable: false, value: this });
        if (value !== undefined)
            (0, runtime_3.reflectionMergePartial)(this, message, value);
        return message;
    }
    internalBinaryRead(reader, length, options, target) {
        return target ?? this.create();
    }
    internalBinaryWrite(message, writer, options) {
        let u = options.writeUnknownFields;
        if (u !== false)
            (u == true ? runtime_2.UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
        return writer;
    }
}
/**
 * @generated MessageType for protobuf message SourceInstallResponse
 */
exports.SourceInstallResponse = new SourceInstallResponse$Type();
// @generated message type with reflection information, may provide speed optimized methods
class SearchData$Type extends runtime_5.MessageType {
    constructor() {
        super("SearchData", [
            { no: 1, name: "query", kind: "scalar", opt: true, T: 9 /*ScalarType.STRING*/ },
            { no: 2, name: "includedTags", kind: "scalar", repeat: 2 /*RepeatType.UNPACKED*/, T: 9 /*ScalarType.STRING*/ },
            { no: 3, name: "excludedTags", kind: "scalar", repeat: 2 /*RepeatType.UNPACKED*/, T: 9 /*ScalarType.STRING*/ },
            { no: 4, name: "itemIndex", kind: "scalar", opt: true, T: 9 /*ScalarType.STRING*/ },
            { no: 5, name: "itemId", kind: "scalar", opt: true, T: 9 /*ScalarType.STRING*/ }
        ]);
    }
    create(value) {
        const message = { includedTags: [], excludedTags: [] };
        globalThis.Object.defineProperty(message, runtime_4.MESSAGE_TYPE, { enumerable: false, value: this });
        if (value !== undefined)
            (0, runtime_3.reflectionMergePartial)(this, message, value);
        return message;
    }
    internalBinaryRead(reader, length, options, target) {
        let message = target ?? this.create(), end = reader.pos + length;
        while (reader.pos < end) {
            let [fieldNo, wireType] = reader.tag();
            switch (fieldNo) {
                case /* optional string query */ 1:
                    message.query = reader.string();
                    break;
                case /* repeated string includedTags */ 2:
                    message.includedTags.push(reader.string());
                    break;
                case /* repeated string excludedTags */ 3:
                    message.excludedTags.push(reader.string());
                    break;
                case /* optional string itemIndex */ 4:
                    message.itemIndex = reader.string();
                    break;
                case /* optional string itemId */ 5:
                    message.itemId = reader.string();
                    break;
                default:
                    let u = options.readUnknownField;
                    if (u === "throw")
                        throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
                    let d = reader.skip(wireType);
                    if (u !== false)
                        (u === true ? runtime_2.UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
            }
        }
        return message;
    }
    internalBinaryWrite(message, writer, options) {
        /* optional string query = 1; */
        if (message.query !== undefined)
            writer.tag(1, runtime_1.WireType.LengthDelimited).string(message.query);
        /* repeated string includedTags = 2; */
        for (let i = 0; i < message.includedTags.length; i++)
            writer.tag(2, runtime_1.WireType.LengthDelimited).string(message.includedTags[i]);
        /* repeated string excludedTags = 3; */
        for (let i = 0; i < message.excludedTags.length; i++)
            writer.tag(3, runtime_1.WireType.LengthDelimited).string(message.excludedTags[i]);
        /* optional string itemIndex = 4; */
        if (message.itemIndex !== undefined)
            writer.tag(4, runtime_1.WireType.LengthDelimited).string(message.itemIndex);
        /* optional string itemId = 5; */
        if (message.itemId !== undefined)
            writer.tag(5, runtime_1.WireType.LengthDelimited).string(message.itemId);
        let u = options.writeUnknownFields;
        if (u !== false)
            (u == true ? runtime_2.UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
        return writer;
    }
}
/**
 * @generated MessageType for protobuf message SearchData
 */
exports.SearchData = new SearchData$Type();
// @generated message type with reflection information, may provide speed optimized methods
class TestData$Type extends runtime_5.MessageType {
    constructor() {
        super("TestData", [
            { no: 1, name: "mangaId", kind: "scalar", opt: true, T: 9 /*ScalarType.STRING*/ },
            { no: 2, name: "chapterId", kind: "scalar", opt: true, T: 9 /*ScalarType.STRING*/ },
            { no: 3, name: "updateTime", kind: "message", T: () => timestamp_1.Timestamp },
            { no: 4, name: "searchData", kind: "message", T: () => exports.SearchData }
        ]);
    }
    create(value) {
        const message = {};
        globalThis.Object.defineProperty(message, runtime_4.MESSAGE_TYPE, { enumerable: false, value: this });
        if (value !== undefined)
            (0, runtime_3.reflectionMergePartial)(this, message, value);
        return message;
    }
    internalBinaryRead(reader, length, options, target) {
        let message = target ?? this.create(), end = reader.pos + length;
        while (reader.pos < end) {
            let [fieldNo, wireType] = reader.tag();
            switch (fieldNo) {
                case /* optional string mangaId */ 1:
                    message.mangaId = reader.string();
                    break;
                case /* optional string chapterId */ 2:
                    message.chapterId = reader.string();
                    break;
                case /* optional google.protobuf.Timestamp updateTime */ 3:
                    message.updateTime = timestamp_1.Timestamp.internalBinaryRead(reader, reader.uint32(), options, message.updateTime);
                    break;
                case /* optional SearchData searchData */ 4:
                    message.searchData = exports.SearchData.internalBinaryRead(reader, reader.uint32(), options, message.searchData);
                    break;
                default:
                    let u = options.readUnknownField;
                    if (u === "throw")
                        throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
                    let d = reader.skip(wireType);
                    if (u !== false)
                        (u === true ? runtime_2.UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
            }
        }
        return message;
    }
    internalBinaryWrite(message, writer, options) {
        /* optional string mangaId = 1; */
        if (message.mangaId !== undefined)
            writer.tag(1, runtime_1.WireType.LengthDelimited).string(message.mangaId);
        /* optional string chapterId = 2; */
        if (message.chapterId !== undefined)
            writer.tag(2, runtime_1.WireType.LengthDelimited).string(message.chapterId);
        /* optional google.protobuf.Timestamp updateTime = 3; */
        if (message.updateTime)
            timestamp_1.Timestamp.internalBinaryWrite(message.updateTime, writer.tag(3, runtime_1.WireType.LengthDelimited).fork(), options).join();
        /* optional SearchData searchData = 4; */
        if (message.searchData)
            exports.SearchData.internalBinaryWrite(message.searchData, writer.tag(4, runtime_1.WireType.LengthDelimited).fork(), options).join();
        let u = options.writeUnknownFields;
        if (u !== false)
            (u == true ? runtime_2.UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
        return writer;
    }
}
/**
 * @generated MessageType for protobuf message TestData
 */
exports.TestData = new TestData$Type();
// @generated message type with reflection information, may provide speed optimized methods
class SourceTestRequest$Type extends runtime_5.MessageType {
    constructor() {
        super("SourceTestRequest", [
            { no: 1, name: "sourceId", kind: "scalar", T: 9 /*ScalarType.STRING*/ },
            { no: 2, name: "data", kind: "message", T: () => exports.TestData }
        ]);
    }
    create(value) {
        const message = { sourceId: "" };
        globalThis.Object.defineProperty(message, runtime_4.MESSAGE_TYPE, { enumerable: false, value: this });
        if (value !== undefined)
            (0, runtime_3.reflectionMergePartial)(this, message, value);
        return message;
    }
    internalBinaryRead(reader, length, options, target) {
        let message = target ?? this.create(), end = reader.pos + length;
        while (reader.pos < end) {
            let [fieldNo, wireType] = reader.tag();
            switch (fieldNo) {
                case /* string sourceId */ 1:
                    message.sourceId = reader.string();
                    break;
                case /* TestData data */ 2:
                    message.data = exports.TestData.internalBinaryRead(reader, reader.uint32(), options, message.data);
                    break;
                default:
                    let u = options.readUnknownField;
                    if (u === "throw")
                        throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
                    let d = reader.skip(wireType);
                    if (u !== false)
                        (u === true ? runtime_2.UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
            }
        }
        return message;
    }
    internalBinaryWrite(message, writer, options) {
        /* string sourceId = 1; */
        if (message.sourceId !== "")
            writer.tag(1, runtime_1.WireType.LengthDelimited).string(message.sourceId);
        /* TestData data = 2; */
        if (message.data)
            exports.TestData.internalBinaryWrite(message.data, writer.tag(2, runtime_1.WireType.LengthDelimited).fork(), options).join();
        let u = options.writeUnknownFields;
        if (u !== false)
            (u == true ? runtime_2.UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
        return writer;
    }
}
/**
 * @generated MessageType for protobuf message SourceTestRequest
 */
exports.SourceTestRequest = new SourceTestRequest$Type();
// @generated message type with reflection information, may provide speed optimized methods
class SourceTestResponse$Type extends runtime_5.MessageType {
    constructor() {
        super("SourceTestResponse", [
            { no: 1, name: "testCase", kind: "scalar", T: 9 /*ScalarType.STRING*/ },
            { no: 2, name: "completeTime", kind: "scalar", T: 4 /*ScalarType.UINT64*/, L: 0 /*LongType.BIGINT*/ },
            { no: 4, name: "failures", kind: "scalar", repeat: 2 /*RepeatType.UNPACKED*/, T: 9 /*ScalarType.STRING*/ }
        ]);
    }
    create(value) {
        const message = { testCase: "", completeTime: 0n, failures: [] };
        globalThis.Object.defineProperty(message, runtime_4.MESSAGE_TYPE, { enumerable: false, value: this });
        if (value !== undefined)
            (0, runtime_3.reflectionMergePartial)(this, message, value);
        return message;
    }
    internalBinaryRead(reader, length, options, target) {
        let message = target ?? this.create(), end = reader.pos + length;
        while (reader.pos < end) {
            let [fieldNo, wireType] = reader.tag();
            switch (fieldNo) {
                case /* string testCase */ 1:
                    message.testCase = reader.string();
                    break;
                case /* uint64 completeTime */ 2:
                    message.completeTime = reader.uint64().toBigInt();
                    break;
                case /* repeated string failures */ 4:
                    message.failures.push(reader.string());
                    break;
                default:
                    let u = options.readUnknownField;
                    if (u === "throw")
                        throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
                    let d = reader.skip(wireType);
                    if (u !== false)
                        (u === true ? runtime_2.UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
            }
        }
        return message;
    }
    internalBinaryWrite(message, writer, options) {
        /* string testCase = 1; */
        if (message.testCase !== "")
            writer.tag(1, runtime_1.WireType.LengthDelimited).string(message.testCase);
        /* uint64 completeTime = 2; */
        if (message.completeTime !== 0n)
            writer.tag(2, runtime_1.WireType.Varint).uint64(message.completeTime);
        /* repeated string failures = 4; */
        for (let i = 0; i < message.failures.length; i++)
            writer.tag(4, runtime_1.WireType.LengthDelimited).string(message.failures[i]);
        let u = options.writeUnknownFields;
        if (u !== false)
            (u == true ? runtime_2.UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
        return writer;
    }
}
/**
 * @generated MessageType for protobuf message SourceTestResponse
 */
exports.SourceTestResponse = new SourceTestResponse$Type();
/**
 * @generated ServiceType for protobuf service PaperbackSourceTester
 */
exports.PaperbackSourceTester = new runtime_rpc_1.ServiceType("PaperbackSourceTester", [
    { name: "installSource", options: {}, I: exports.SourceInstallRequest, O: exports.SourceInstallResponse },
    { name: "testSource", serverStreaming: true, options: {}, I: exports.SourceTestRequest, O: exports.SourceTestResponse }
]);
