// @paperback/toolchain/dist/commands/migrate.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
/* eslint-disable unicorn/no-useless-undefined */
const node_fs_1 = tslib_1.__importDefault(require("node:fs"));
const node_path_1 = tslib_1.__importDefault(require("node:path"));
const typescript_1 = tslib_1.__importDefault(require("typescript"));
const command_1 = require("../command");
const shelljs_1 = tslib_1.__importDefault(require("shelljs"));
class Migrate extends command_1.CLICommand {
    async run() {
        this.log('Traversing /src directory');
        // eslint-disable-next-line unicorn/prefer-module
        const myPkgPath = node_path_1.default.join(__dirname, '../../package.json');
        const myPkg = JSON.parse(node_fs_1.default.readFileSync(myPkgPath).toString());
        const cwd = process.cwd();
        const srcDir = node_path_1.default.join(cwd, 'src');
        const packagePath = node_path_1.default.join(cwd, 'package.json');
        if (!node_fs_1.default.existsSync(packagePath)) {
            throw new Error('could not locate package.json');
        }
        if (!node_fs_1.default.lstatSync(srcDir).isDirectory()) {
            throw new Error('/src is not a directory');
        }
        const tsFiles = [];
        this.goThrough(srcDir, tsFiles);
        this.migrateTypescriptFiles(tsFiles);
        const pkg = JSON.parse(node_fs_1.default.readFileSync(packagePath).toString());
        delete pkg.dependencies['paperback-extensions-common'];
        delete pkg.dependencies['paperback-cli'];
        this.log(`setting @paperback/types and @paperback/toolchain to ${myPkg.version}`);
        pkg.dependencies['@paperback/types'] = myPkg.version;
        pkg.dependencies['@paperback/toolchain'] = myPkg.version;
        node_fs_1.default.writeFileSync(packagePath, JSON.stringify(pkg, undefined, 2));
        shelljs_1.default.exec('npm i');
    }
    goThrough(dir, tsFiles) {
        for (const file of node_fs_1.default.readdirSync(dir)) {
            const filePath = node_path_1.default.join(dir, file);
            if (node_fs_1.default.lstatSync(filePath).isDirectory()) {
                this.goThrough(filePath, tsFiles);
                continue;
            }
            if (!file.endsWith('.ts')) {
                continue;
            }
            tsFiles.push(filePath);
        }
    }
    migrateTypescriptFiles(tsFiles) {
        const prog = typescript_1.default.createProgram(tsFiles, {});
        const printer = typescript_1.default.createPrinter({ newLine: typescript_1.default.NewLineKind.LineFeed, omitTrailingSemicolon: true });
        const tsFilesSet = new Set(tsFiles);
        for (const sourceFile of prog.getSourceFiles()) {
            const filePath = node_path_1.default.normalize(sourceFile.fileName);
            if (!tsFilesSet.has(filePath))
                continue;
            this.log('PROCESSING ' + filePath);
            const transformer = new FileTransformer(sourceFile);
            const transformedFiles = transformer.transform().transformed;
            if (transformer.changes.length > 0) {
                this.log(' - ' + transformer.changes.join('\n - '));
                node_fs_1.default.writeFileSync(filePath, printer.printList(typescript_1.default.ListFormat.MultiLine, typescript_1.default.factory.createNodeArray(transformedFiles), sourceFile));
            }
        }
    }
}
exports.default = Migrate;
Migrate.description = 'Migrate 0.7 sources to 0.8';
Migrate.examples = [
    '<%= config.bin %> <%= command.id %>',
];
class FileTransformer {
    constructor(sourceFile) {
        this.sourceFile = sourceFile;
        this.changes = [];
    }
    getPropertyAssignmentExpressions(node, obj) {
        if (!typescript_1.default.isObjectLiteralExpression(node))
            return obj;
        const keySet = new Set(Object.keys(obj));
        for (const prop of node.properties) {
            if (!prop.name || !typescript_1.default.isIdentifier(prop.name) || !keySet.has(prop.name.text))
                continue;
            if (prop && typescript_1.default.isPropertyAssignment(prop)) {
                // eslint-disable-next-line @typescript-eslint/ban-ts-comment
                // @ts-ignore
                obj[prop.name.text] = prop.initializer;
            }
            if (prop && typescript_1.default.isShorthandPropertyAssignment(prop)) {
                // eslint-disable-next-line @typescript-eslint/ban-ts-comment
                // @ts-ignore
                obj[prop.name.text] = prop.name;
            }
        }
        return obj;
    }
    getRenamedTypeWrapper(callExp) {
        const expression = callExp.expression;
        if (!typescript_1.default.isIdentifier(expression) || !expression.text.startsWith('create'))
            return callExp;
        this.changes.push(`Updating creation wrapper call ${expression.text} to use the App namespace`);
        switch (expression.text) {
            case 'createHomeSection': {
                const propAssignments = this.getPropertyAssignmentExpressions(callExp.arguments[0], {
                    // eslint-disable-next-line camelcase
                    view_more: undefined,
                    type: undefined,
                });
                return typescript_1.default.factory.createCallExpression(typescript_1.default.factory.createPropertyAccessExpression(typescript_1.default.factory.createIdentifier('App'), typescript_1.default.factory.createIdentifier('createHomeSection')), callExp.typeArguments, [typescript_1.default.factory.createObjectLiteralExpression([
                        ...callExp.arguments[0].properties.filter(x => {
                            const removedProps = new Set(Object.keys(propAssignments));
                            if (x.name && typescript_1.default.isIdentifier(x.name) && removedProps.has(x.name.text))
                                return false;
                            return true;
                        }),
                        typescript_1.default.factory.createPropertyAssignment(typescript_1.default.factory.createIdentifier('containsMoreItems'), propAssignments.view_more ?? typescript_1.default.factory.createIdentifier('false')),
                        typescript_1.default.factory.createPropertyAssignment(typescript_1.default.factory.createIdentifier('type'), propAssignments.type ?? typescript_1.default.factory.createStringLiteral('singleRowNormal')),
                    ], true)]);
            }
            case 'createChapterDetails':
                return typescript_1.default.factory.createCallExpression(typescript_1.default.factory.createPropertyAccessExpression(typescript_1.default.factory.createIdentifier('App'), typescript_1.default.factory.createIdentifier('createChapterDetails')), callExp.typeArguments, [typescript_1.default.factory.createObjectLiteralExpression([
                        ...callExp.arguments[0].properties.filter(x => {
                            const removedProps = new Set(['longStrip']);
                            if (x.name && typescript_1.default.isIdentifier(x.name) && removedProps.has(x.name.text))
                                return false;
                            return true;
                        }),
                    ], true)]);
            case 'createChapter': {
                return typescript_1.default.factory.createCallExpression(typescript_1.default.factory.createPropertyAccessExpression(typescript_1.default.factory.createIdentifier('App'), typescript_1.default.factory.createIdentifier('createChapter')), callExp.typeArguments, [typescript_1.default.factory.createObjectLiteralExpression([
                        ...callExp.arguments[0].properties.filter(x => {
                            const removedProps = new Set(['mangaId']);
                            if (x.name && typescript_1.default.isIdentifier(x.name) && removedProps.has(x.name.text))
                                return false;
                            return true;
                        }),
                    ], true)]);
            }
            case 'createIconText': {
                const propAssignments = this.getPropertyAssignmentExpressions(callExp.arguments[0], { text: undefined });
                return propAssignments.text;
            }
            case 'createManga': {
                const propAssignments = this.getPropertyAssignmentExpressions(callExp.arguments[0], { id: undefined });
                return typescript_1.default.factory.createCallExpression(typescript_1.default.factory.createPropertyAccessExpression(typescript_1.default.factory.createIdentifier('App'), typescript_1.default.factory.createIdentifier('createSourceManga')), callExp.typeArguments, [typescript_1.default.factory.createObjectLiteralExpression([
                        typescript_1.default.factory.createPropertyAssignment(typescript_1.default.factory.createIdentifier('id'), propAssignments.id ?? typescript_1.default.factory.createIdentifier('undefined')),
                        typescript_1.default.factory.createPropertyAssignment(typescript_1.default.factory.createIdentifier('mangaInfo'), typescript_1.default.factory.createCallExpression(typescript_1.default.factory.createPropertyAccessExpression(typescript_1.default.factory.createIdentifier('App'), typescript_1.default.factory.createIdentifier('createMangaInfo')), callExp.typeArguments, [typescript_1.default.factory.createObjectLiteralExpression([
                                ...callExp.arguments[0].properties.filter(x => {
                                    const removedProps = new Set(Object.keys(propAssignments));
                                    if (x.name && typescript_1.default.isIdentifier(x.name) && removedProps.has(x.name.text))
                                        return false;
                                    return true;
                                }),
                            ], true)])),
                    ], true)]);
            }
            case 'createMangaTile': {
                const propAssignments = this.getPropertyAssignmentExpressions(callExp.arguments[0], {
                    id: undefined,
                    subtitleText: undefined,
                    subtitle: undefined,
                });
                return typescript_1.default.factory.createCallExpression(typescript_1.default.factory.createPropertyAccessExpression(typescript_1.default.factory.createIdentifier('App'), typescript_1.default.factory.createIdentifier('createPartialSourceManga')), callExp.typeArguments, [typescript_1.default.factory.createObjectLiteralExpression([
                        ...callExp.arguments[0].properties.filter(x => {
                            const removedProps = new Set(Object.keys(propAssignments));
                            if (x.name && typescript_1.default.isIdentifier(x.name) && removedProps.has(x.name.text))
                                return false;
                            return true;
                        }),
                        typescript_1.default.factory.createPropertyAssignment(typescript_1.default.factory.createIdentifier('mangaId'), propAssignments.id ?? typescript_1.default.factory.createIdentifier('undefined')),
                        typescript_1.default.factory.createPropertyAssignment(typescript_1.default.factory.createIdentifier('subtitle'), propAssignments.subtitle ??
                            propAssignments.subtitleText ??
                            typescript_1.default.factory.createIdentifier('undefined')),
                    ], true)]);
            }
            case 'createRequestObject':
                return typescript_1.default.factory.createCallExpression(typescript_1.default.factory.createPropertyAccessExpression(typescript_1.default.factory.createIdentifier('App'), typescript_1.default.factory.createIdentifier('createRequest')), callExp.typeArguments, callExp.arguments);
            default: break;
        }
        return typescript_1.default.factory.createCallExpression(typescript_1.default.factory.createPropertyAccessExpression(typescript_1.default.factory.createIdentifier('App'), typescript_1.default.factory.createIdentifier(expression.text)), callExp.typeArguments, callExp.arguments);
    }
    getRenamedImport(namedImport) {
        switch (namedImport.name.text) {
            case 'Manga': return typescript_1.default.factory.createImportSpecifier(false, undefined, typescript_1.default.factory.createIdentifier('SourceManga'));
            case 'TagType': return typescript_1.default.factory.createImportSpecifier(false, undefined, typescript_1.default.factory.createIdentifier('BadgeColor'));
            case 'MangaTile': return typescript_1.default.factory.createImportSpecifier(false, undefined, typescript_1.default.factory.createIdentifier('PartialSourceManga'));
            default:
                return namedImport;
        }
    }
    getRenamedType(type) {
        if (!typescript_1.default.isIdentifier(type.typeName))
            return type;
        switch (type.typeName.text) {
            case 'Manga':
                return typescript_1.default.factory.createTypeReferenceNode(typescript_1.default.factory.createIdentifier('SourceManga'), undefined);
            case 'MangaTile':
                return typescript_1.default.factory.createTypeReferenceNode(typescript_1.default.factory.createIdentifier('PartialSourceManga'), undefined);
        }
        return type;
    }
    getRenamedPropertyAccessor(property) {
        const expression = property.expression;
        if (!typescript_1.default.isIdentifier(expression))
            return property;
        switch (expression.text) {
            case 'TagType':
                return typescript_1.default.factory.createPropertyAccessExpression(typescript_1.default.factory.createIdentifier('BadgeColor'), property.name);
        }
        return property;
    }
    transform() {
        return typescript_1.default.transform(this.sourceFile, [this.transformerFactory.bind(this)]);
    }
    transformerFactory(context) {
        const changes = this.changes;
        return (rootNode) => {
            const visitNode = (node) => {
                // Migrate the import declerations
                if (typescript_1.default.isImportDeclaration(node)) {
                    const moduleSpecifier = node.moduleSpecifier;
                    if (typescript_1.default.isStringLiteral(moduleSpecifier) && moduleSpecifier.text.startsWith('paperback-extensions-common')) {
                        changes.push('Import to @paperback/types');
                        const namedImports = node.importClause?.namedBindings;
                        return typescript_1.default.visitEachChild(typescript_1.default.factory.createImportDeclaration(undefined, node.modifiers, typescript_1.default.factory.createImportClause(node.importClause?.isTypeOnly ?? false, node.importClause?.name, namedImports?.elements ? typescript_1.default.factory.createNamedImports(namedImports.elements.map(x => this.getRenamedImport(x))) : node.importClause?.namedBindings), typescript_1.default.factory.createStringLiteral('@paperback/types'), node.assertClause), visitNode.bind(this), context);
                    }
                }
                if (typescript_1.default.isCallExpression(node)) {
                    const expression = node.expression;
                    if (typescript_1.default.isIdentifier(expression)) {
                        return typescript_1.default.visitEachChild(this.getRenamedTypeWrapper(node), visitNode.bind(this), context);
                    }
                }
                if (typescript_1.default.isTypeReferenceNode(node)) {
                    return typescript_1.default.visitEachChild(this.getRenamedType(node), visitNode.bind(this), context);
                }
                if (typescript_1.default.isPropertyAccessExpression(node)) {
                    return typescript_1.default.visitEachChild(this.getRenamedPropertyAccessor(node), visitNode.bind(this), context);
                }
                return typescript_1.default.visitEachChild(node, visitNode.bind(this), context);
            };
            return typescript_1.default.visitNode(rootNode, visitNode.bind(this));
        };
    }
}
