// @paperback/toolchain/dist/commands/serve.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const core_1 = require("@oclif/core");
const command_1 = require("../command");
const bundle_1 = tslib_1.__importDefault(require("./bundle"));
const server_1 = tslib_1.__importDefault(require("../server"));
const utils_1 = tslib_1.__importDefault(require("../utils"));
const chalk_1 = tslib_1.__importDefault(require("chalk"));
const node_fs_1 = require("node:fs");
const path = tslib_1.__importStar(require("node:path"));
const cli = core_1.CliUx.ux;
class Serve extends command_1.CLICommand {
    constructor() {
        super(...arguments);
        this.workingPath = process.cwd();
    }
    async _bundleRun(useNodeFS) {
        await bundle_1.default.run([
            ...(useNodeFS ? ['--use-node-fs'] : []),
        ]);
    }
    async _rebuildSources(useNodeFS) {
        this.log(chalk_1.default.underline.blue('Building Sources'));
        await this._bundleRun(useNodeFS);
        this.log(chalk_1.default.underline.blue('Sources Rebuilt'));
    }
    async run() {
        const { flags } = await this.parse(Serve);
        console.clear();
        this.log(chalk_1.default.underline.blue('Building Sources'));
        // Make sure the repo is bundled
        await this._bundleRun(flags['use-node-fs']);
        this.log();
        this.log(chalk_1.default.underline.blue('Starting Server on port ' + flags.port));
        const server = new server_1.default(flags.port);
        server.start();
        this.log();
        this.log((0, chalk_1.default) `For a list of commands do {green h} or {green help}`);
        let isRebuildingSources = false;
        let watcher;
        if (flags.watch) {
            watcher = (0, node_fs_1.watch)(path.join(this.workingPath, 'src'), { recursive: true }, async (_eventType, filename) => {
                if (isRebuildingSources === false && filename?.endsWith('.ts')) {
                    isRebuildingSources = true;
                    await this._rebuildSources(flags['use-node-fs']);
                    isRebuildingSources = false;
                }
            });
        }
        let stopServer = false;
        while (!stopServer) {
            // eslint-disable-next-line no-await-in-loop
            const input = (await cli.prompt(utils_1.default.prefixTime(''), { required: false }))?.trim() ?? '';
            if (input === 'h' || input === 'help') {
                this.log(chalk_1.default.underline.bold('Help'));
                this.log('  h, help - Display this message');
                this.log('  s, stop - Stop the server');
                this.log('  r, rebuild - Rebuilds the sources');
                this.log('  rs, restart - Restart the server, also rebuilding the sources');
            }
            if (input === 's' || input === 'stop') {
                watcher?.close();
                stopServer = true;
            }
            if (input === 'rs' || input === 'restart') {
                server.stop();
                console.clear();
                this.log(chalk_1.default.underline.blue('Building Sources'));
                // Make sure the repo is bundled
                // eslint-disable-next-line no-await-in-loop
                await this._bundleRun(flags['use-node-fs']);
                this.log();
                this.log(chalk_1.default.underline.blue('Starting Server on port ' + flags.port));
                server.start();
                this.log();
                this.log((0, chalk_1.default) `For a list of commands do {green h} or {green help}`);
            }
            if (input === 'r' || input === 'rebuild') {
                console.clear();
                // eslint-disable-next-line no-await-in-loop
                await this._bundleRun(flags['use-node-fs']);
                this.log();
                this.log((0, chalk_1.default) `For a list of commands do {green h} or {green help}`);
            }
        }
        // eslint-disable-next-line no-process-exit, unicorn/no-process-exit
        process.exit(0);
    }
}
exports.default = Serve;
Serve.description = 'Build the sources and start a local server';
Serve.flags = {
    help: core_1.Flags.help({ char: 'h' }),
    port: core_1.Flags.integer({ char: 'p', default: 8080 }),
    watch: core_1.Flags.boolean({ char: 'w', description: 'Rebuild sources on typescript file changes within directory', default: false }),
    'use-node-fs': core_1.Flags.boolean({ description: 'For more info, check https://github.com/Paperback-iOS/paperback-toolchain/pull/4#issuecomment-1791566399', required: false, default: false }),
};
