// @paperback/toolchain/dist/commands/test.js
"use strict";
/* eslint-disable no-await-in-loop */
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const core_1 = require("@oclif/core");
const command_1 = require("../command");
const bundle_1 = tslib_1.__importDefault(require("./bundle"));
const path = tslib_1.__importStar(require("node:path"));
const fs = tslib_1.__importStar(require("node:fs"));
const chalk_1 = tslib_1.__importDefault(require("chalk"));
const source_tester_1 = require("../source-tester");
const shelljs_1 = tslib_1.__importDefault(require("shelljs"));
const utils_1 = tslib_1.__importDefault(require("../utils"));
const server_1 = tslib_1.__importDefault(require("../server"));
const PDTSourceTester_grpc_client_1 = require("../devtools/generated/typescript/PDTSourceTester.grpc-client");
const grpc_js_1 = require("@grpc/grpc-js");
const ip_1 = tslib_1.__importDefault(require("ip"));
class Test extends command_1.CLICommand {
    constructor() {
        super(...arguments);
        this.utils = undefined;
    }
    async run() {
        const { flags, args } = await this.parse(Test);
        this.utils = flags['use-node-fs'] ? new utils_1.default(false) : new utils_1.default(true);
        const cwd = process.cwd();
        const sourceId = args.source;
        let sourcesDirPath;
        let testClient;
        let installClient;
        if (flags.ip) {
            const grpcClient = new PDTSourceTester_grpc_client_1.PaperbackSourceTesterClient(`${flags.ip}:${flags.port}`, grpc_js_1.credentials.createInsecure());
            sourcesDirPath = path.join(cwd, 'bundles');
            const tester = new source_tester_1.OnDeviceSourceTester(grpcClient);
            testClient = tester;
            installClient = tester;
        }
        else {
            this.log(`\n${chalk_1.default.bold.underline.bgBlue.white('Transpiling Sources')}`);
            sourcesDirPath = path.join(cwd, 'tmp');
            await this.measure('Time', utils_1.default.headingFormat, async () => {
                await this.utils.deleteFolderRecursive(sourcesDirPath);
                shelljs_1.default.exec('npx tsc --outDir tmp');
            });
            testClient = new source_tester_1.SourceTester(sourcesDirPath);
        }
        const sourcesToTest = this.getSourceIdsToTest(sourceId, sourcesDirPath);
        if (installClient) {
            await this.installSources(sourcesToTest, installClient, flags['use-node-fs']);
        }
        await this.testSources(sourcesToTest, testClient);
    }
    async logSourceTestReponse(response) {
        this.log(`${chalk_1.default.red.bold('#')} ${chalk_1.default.bold(response.testCase)}: ${chalk_1.default.green(response.completeTime + 'ms')}`);
        const failures = response.failures;
        for (const failure of failures) {
            this.log(`- ${chalk_1.default.white.bgRed('[FAILURE]')} ${failure}`);
        }
        if (failures.length > 0)
            this.log();
    }
    getSourceIdsToTest(sourceId, bundleDir) {
        let sourcesToTest = [];
        sourcesToTest = sourceId ? fs.readdirSync(bundleDir).filter(file => {
            return file.toLowerCase() === sourceId.toLowerCase() && fs.statSync(path.join(bundleDir, file)).isDirectory();
        }) : fs.readdirSync(bundleDir).filter(file => fs.statSync(path.join(bundleDir, file)).isDirectory());
        return sourcesToTest;
    }
    async testSources(sources, client) {
        for (const sourceId of sources) {
            this.log(`\n${chalk_1.default.bold.underline.bgBlue.white(`Testing ${sourceId}`)}`);
            const request = { sourceId };
            const dataPath = path.join(process.cwd(), 'src', sourceId, `${sourceId}.test.json`);
            if (fs.existsSync(dataPath)) {
                request.data = JSON.parse(fs.readFileSync(dataPath, { encoding: 'utf8' }));
            }
            await client.testSource(request, this.logSourceTestReponse.bind(this));
        }
    }
    async installSources(sources, client, useNodeFS) {
        if (!client.installSource) {
            return;
        }
        await bundle_1.default.run([
            ...(useNodeFS ? ['--use-node-fs'] : []),
        ]);
        const server = new server_1.default(8000);
        server.start();
        // Install the sources that we need to test on the App
        for (const source of sources) {
            this.log(`\n${chalk_1.default.bold.underline.bgBlue.white(`Installing ${source}`)}`);
            // Make sure the source is installed on the app
            await client.installSource({ sourceId: source, repoBaseUrl: `http://${ip_1.default.address()}:${server.port}` });
        }
        // Sleep before stopping the server because the installSource returns
        // immediately because the install is queued and processed asynchronously by the app.
        await sleep(3000);
        server.stop();
    }
}
exports.default = Test;
Test.description = 'describe the command here';
Test.flags = {
    ip: core_1.Flags.string({ name: 'ip', default: undefined }),
    port: core_1.Flags.integer({ name: 'port', default: 27015 }),
    'use-node-fs': core_1.Flags.boolean({ description: 'For more info, check https://github.com/Paperback-iOS/paperback-toolchain/pull/4#issuecomment-1791566399', required: false }),
};
Test.args = [
    {
        name: 'source',
        required: false,
        description: '(optional) The source to test',
        default: undefined,
    },
];
function sleep(ms) {
    // eslint-disable-next-line no-promise-executor-return
    return new Promise(resolve => setTimeout(resolve, ms));
}
