// @paperback/toolchain/dist/expect.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.expect = void 0;
class Expect {
    // eslint-disable-next-line no-useless-constructor
    constructor(value) {
        this.value = value;
        this.predicates = [];
    }
    toExist() {
        this.predicates.push(value => Boolean(value));
        return this;
    }
    toBeNull() {
        this.predicates.push(value => value === null);
        return this;
    }
    toBeUndefined() {
        this.predicates.push(value => value === undefined);
        return this;
    }
    toBeNullOrUndefined() {
        this.predicates.push(value => value === undefined || value === null);
        return this;
    }
    toBeNotNull() {
        this.predicates.push(value => value !== null);
        return this;
    }
    toBeNotUndefined() {
        this.predicates.push(value => value !== undefined);
        return this;
    }
    toBeNotNullOrUndefined() {
        this.predicates.push(value => value !== undefined && value !== null);
        return this;
    }
    toBeEqual(otherValue) {
        this.predicates.push(value => value === otherValue);
        return this;
    }
    toBeNotEqual(otherValue) {
        this.predicates.push(value => value !== otherValue);
        return this;
    }
    toBeGreaterThan(otherValue) {
        this.predicates.push(value => value > otherValue);
        return this;
    }
    toBeGreaterThanOrEqualTo(otherValue) {
        this.predicates.push(value => value >= otherValue);
        return this;
    }
    toBeLessThan(otherValue) {
        this.predicates.push(value => value < otherValue);
        return this;
    }
    toBeLessThanOrEqualTo(otherValue) {
        this.predicates.push(value => value <= otherValue);
        return this;
    }
    toMatchPredicate(predicate) {
        this.predicates.push(predicate);
        return this;
    }
    toNotMatchPredicate(predicate) {
        this.predicates.push(x => !predicate(x));
        return this;
    }
    assertWithError(error) {
        if (this.predicates === undefined || this.predicates.length === 0) {
            throw new Error('comparison value undefined');
        }
        // eslint-disable-next-line unicorn/no-array-reduce
        return this.predicates.reduce((previousValue, currentValue) => {
            return previousValue && currentValue(this.value);
        }, true) ? undefined : error;
    }
}
class NamedExpect extends Expect {
    constructor(variableName, value) {
        super(value);
        this.variableName = variableName;
    }
}
function expect(value) {
    return new Expect(value);
}
exports.expect = expect;
