// @paperback/toolchain/dist/source-tester.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OnDeviceSourceTester = exports.SourceTester = void 0;
const tslib_1 = require("tslib");
require("@paperback/runtime-polyfills");
const path = tslib_1.__importStar(require("node:path"));
const cheerio_1 = tslib_1.__importDefault(require("cheerio"));
const node_fs_1 = tslib_1.__importDefault(require("node:fs"));
const chai_1 = require("chai");
class SourceTester {
    // eslint-disable-next-line no-useless-constructor
    constructor(bundleDir) {
        this.bundleDir = bundleDir;
    }
    async testSource(request, callback) {
        const testData = request.data ?? {};
        const sourceFilePath = path.join(this.bundleDir, request.sourceId, request.sourceId);
        const sourceFile = require(sourceFilePath);
        const source = new sourceFile[request.sourceId](cheerio_1.default);
        const testCase = async (testCase, runner) => {
            try {
                callback({
                    testCase,
                    failures: [],
                    completeTime: BigInt(await this.measureTime(async () => {
                        await runner();
                    })),
                });
            }
            catch (error) {
                if (typeof error === 'string') {
                    callback({ testCase, completeTime: 0n, failures: [error] });
                }
                else if (typeof error === 'object' && error.message) {
                    callback({ testCase, completeTime: 0n, failures: [error.message] });
                }
            }
        };
        const customRunnerPath = path.join(process.cwd(), 'src', request.sourceId, `${request.sourceId}.test.js`);
        if (node_fs_1.default.existsSync(customRunnerPath)) {
            const customRunner = require(customRunnerPath);
            if (customRunner.runTests) {
                await customRunner.runTests(testCase, testData, source, this.runTests.bind(this));
                return;
            }
        }
        this.runTests(testCase, testData, source);
    }
    async runTests(testCase, testData, source) {
        const homepageProvider = source;
        if (homepageProvider?.getHomePageSections) {
            await testCase('source should return homepage sections containing items', async () => {
                const sections = {};
                await homepageProvider.getHomePageSections(section => {
                    sections[section.id] = section;
                });
                const sectionKeys = Object.keys(sections);
                for (const key of sectionKeys) {
                    (0, chai_1.expect)(sections[key].items, `section[${key}] has no items`).to.not.be.empty;
                    if (!testData.mangaId) {
                        testData.mangaId = sections[key].items[0].mangaId;
                    }
                    if (!testData.searchData) {
                        testData.searchData = {
                            query: sections[key].items[0].title,
                            excludedTags: [],
                            includedTags: [],
                        };
                    }
                }
            });
        }
        const mangaProvider = source;
        if (mangaProvider?.getMangaDetails) {
            await testCase('source should return valid SourceManga definition', async () => {
                (0, chai_1.expect)(testData.mangaId, 'no mangaId provided and could not infer it from homepage').to.exist;
                const sourceManga = await mangaProvider.getMangaDetails(testData.mangaId);
                (0, chai_1.expect)(sourceManga.id, 'returned manga id does not match the queried manga id').to.equal(testData.mangaId);
                (0, chai_1.expect)(sourceManga.mangaInfo, 'MangaInfo does not exist').to.exist;
            });
        }
        const chapterProvider = source;
        if (chapterProvider?.getChapters) {
            await testCase('source should return some chapters', async () => {
                (0, chai_1.expect)(testData.mangaId, 'no mangaId provided and could not infer it from homepage').to.exist;
                const chapters = await chapterProvider.getChapters(testData.mangaId);
                (0, chai_1.expect)(chapters, `one or more chapters should be returned for manga '${testData.mangaId}'`).to.not.be.empty;
                if (!testData.chapterId) {
                    testData.chapterId = chapters[0].id;
                }
            });
        }
        if (chapterProvider?.getChapterDetails) {
            await testCase('source should return chapter details with some pages', async () => {
                (0, chai_1.expect)(testData.mangaId, 'no mangaId provided and could not infer it from homepage').to.exist;
                (0, chai_1.expect)(testData.chapterId, 'no chapterId provided and could not infer it from chapter list').to.exist;
                const chapterDetails = await chapterProvider.getChapterDetails(testData.mangaId, testData.chapterId);
                (0, chai_1.expect)(chapterDetails.id, 'returned chapter id does not match the queried chapter id').to.equal(testData.chapterId);
                (0, chai_1.expect)(chapterDetails.mangaId, 'returned manga id does not match the queried manga id').to.equal(testData.mangaId);
                (0, chai_1.expect)(chapterDetails.pages, 'chapter must have pages').to.not.be.empty;
            });
        }
        const searchResultsProvider = source;
        if (searchResultsProvider?.getSearchResults) {
            await testCase('source should return some search results', async () => {
                (0, chai_1.expect)(testData.searchData, 'no searchData provided and could not infer it from homepage').to.exist;
                const searchResults = await searchResultsProvider?.getSearchResults({
                    title: testData.searchData.query ?? '',
                    excludedTags: testData.searchData.excludedTags?.map(x => ({ id: x, label: x })) ?? [],
                    includedTags: testData.searchData.includedTags?.map(x => ({ id: x, label: x })) ?? [],
                    parameters: {},
                }, {});
                (0, chai_1.expect)(searchResults.results, `one or more search results should be returned for ${JSON.stringify(testData.searchData)}`).to.not.be.empty;
            });
        }
    }
    async measureTime(closure) {
        const startTime = process.hrtime.bigint();
        await closure();
        const endTime = Number(process.hrtime.bigint() - startTime);
        return Math.ceil(endTime / 1000000);
    }
}
exports.SourceTester = SourceTester;
class OnDeviceSourceTester {
    // eslint-disable-next-line no-useless-constructor
    constructor(grpcClient) {
        this.grpcClient = grpcClient;
    }
    async installSource(request) {
        await new Promise((resolve, reject) => {
            this.grpcClient.installSource(request, (error, response) => (error) ?
                reject(error) :
                resolve(response));
        });
    }
    async testSource(request, callback) {
        await new Promise((resolve, reject) => {
            this.grpcClient.testSource(request)
                .on('end', resolve)
                .on('error', reject)
                .on('data', callback);
        });
    }
}
exports.OnDeviceSourceTester = OnDeviceSourceTester;
