// @paperback/toolchain/dist/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
/* eslint-disable unicorn/prefer-module */
const chalk_1 = tslib_1.__importDefault(require("chalk"));
const node_fs_1 = require("node:fs");
class Utils {
    constructor(useFSUtils = true) {
        let canUseFSUtils = false;
        if (useFSUtils) {
            try {
                this.fsutils = require('@ivanmatthew/fs-utils-rs');
                canUseFSUtils = true;
            }
            catch {
                canUseFSUtils = false;
            }
        }
        if (!canUseFSUtils && useFSUtils) {
            const platform = require('node:os').platform();
            Utils.log((0, chalk_1.default) `{yellow WARNING:} {gray (rust) fs-utils} is not available for your platform '${platform}', falling back to {gray node fs}!`);
        }
    }
    static fixedWidth(number, width) {
        return (Array.from({ length: width }).join('0') + number).slice(-width);
    }
    static prefixTime(message = '') {
        const date = new Date();
        const time = `${this.fixedWidth(date.getHours(), 2)}:${this.fixedWidth(date.getMinutes(), 2)}:${this.fixedWidth(date.getSeconds(), 2)}:${this.fixedWidth(date.getMilliseconds(), 4)}`;
        return (0, chalk_1.default) `[{gray ${time}}] ${message}`;
    }
    static log(message = '') {
        const cursorTo = process.stdout.cursorTo;
        if (cursorTo) {
            cursorTo(0);
        }
        process.stdout.write(this.prefixTime(message) + '\n');
    }
    static error(message = '') {
        this.log((0, chalk_1.default) `{red ${message}}`);
    }
    static time(label, template = '$1') {
        const startTime = process.hrtime.bigint();
        return {
            end: () => {
                const hrend = process.hrtime.bigint() - startTime;
                this.log(`${template.replace('$1', label)}: ${chalk_1.default.green((hrend / BigInt(1000000)) + 'ms')}`);
            },
        };
    }
    async deleteFolderRecursive(folderPath) {
        folderPath = folderPath.trim();
        if (folderPath.length === 0 || folderPath === '/')
            return;
        if (this.fsutils) {
            return new Promise((res, rej) => {
                try {
                    this.fsutils.deleteFolderRecursive(folderPath);
                    res();
                }
                catch (error) {
                    rej(error);
                }
            });
        }
        await node_fs_1.promises.rm(folderPath, { recursive: true }).catch(error => {
            console.error(`Error deleting and/or creating bundles directory '${folderPath}}'`, error);
        });
    }
}
exports.default = Utils;
Utils.headingFormat = (0, chalk_1.default) `{bold {red #} $1}`;
